src/library/congr_lemma.cpp

```diff
@@ -424,20 +460,16 @@ struct congr_lemma_manager {
                 // Propositions are all definitionally equal, so we don't need to make this Eq.
                 kinds[i] = congr_arg_kind::Cast;
             } else if (pinfos[i].is_inst_implicit()) {
-                // Instance implicits should be Fixed or Cast as appropriate.
-                if (ssinfos[i].is_subsingleton() && !pinfos[i].has_fwd_deps())
-                    kinds[i] = congr_arg_kind::Cast;
-                else
-                    kinds[i] = congr_arg_kind::Fixed;
-            } else if (ssinfos[i].is_subsingleton()) {
-                // If there are backwards dependencies on Eq arguments, then
-                // we will use subsingleton elimination to prove this
-                // equality.
-                // Otherwise we can let this arg stay Eq without problems.
-                for (auto j : pinfos[i].get_back_deps()) {
-                    if (kinds[j] == congr_arg_kind::Eq) {
-                        kinds[i] = congr_arg_kind::Cast;
-                        break;
+                kinds[i] = congr_arg_kind::Fixed;
+
+                // Upgrade decidability instances which depend on Eq to DecInst.
+                // (Otherwise fix_kinds_for_dependencies will downgrade them to Fixed)
+                if (pinfos[i].is_dec_inst()) {
```

Comment:
As far as I can tell, there is no code specific to `decidable` in the simplifier except for this code that detects whether a parameter is of the form `decidable _`, correct?

Since it's quite likely we want to treat `fintype` exactly the same, I feel the intention is clearer if the `congr_arg_kind` was called `Subsingleton` or something to indicate what we want to do with it, rather than what it looks like.

Reply:
Okay, renamed to `SubsingletonInst`.